Add metrics reference (#2931)
sensu/sensu-docs: content/sensu-go/6.2/observability-pipeline/observe-schedule/metrics.md

---
title: "Metrics reference"
linkTitle: "Metrics Reference"
reference_title: "Metrics"
type: "reference"
description: "Use checks to collect and process service metrics for your infrastructure with the Sensu observability pipeline. Read this reference doc to learn about Sensu Go's first-class support for collecting and processing metrics."
weight: 33
version: "6.2"
product: "Sensu Go"
platformContent: false
menu:
  sensu-go-6.2:
    parent: observe-schedule
---

Sensu Go offers first-class built-in support for collecting and processing metrics for your entire infrastructure.

> In Sensu, metrics are events that contain metrics data.

Reviewer: ```suggestion
In Sensu, metrics are an optional component of event data.
```

> Sensu metrics can be metrics-only events created with data from the [Sensu StatsD listener][2] or [Prometheus][7].

Reviewer: ```suggestion
Sensu events may contain check execution results, metrics, or both. Certain inputs like the [Sensu StatsD listener][2] or patterns like the [Prometheus][7] collector pattern will create metric-only events.
```

They can also be status and metrics events from [check output metric extraction][4].

> 

Author: Maybe we should add a note here to explain that the metrics reference does not describe the same functionality as the [metrics API](https://docs.sensu.io/sensu-go/latest/api/metrics/). I'm not sure if that's a common misunderstanding, but most of our other references coordinate with an API doc so it might be worth mentioning.

Author: Added

## Extract metrics from check output

The Sensu agent can extract metrics data from check command output and populate an event's [metrics attribute][5] before sending the event to the Sensu backend for [processing][11].

To extract metrics from check output, the check must include an [`output_metric_format` attribute][10] that specifies one of Sensu's [supported output metric formats][9].
When a check includes the `output_metric_format` attribute, Sensu will extract the specified metrics from the check output and add them to the event data in the [metrics attribute][5].

> In addition, the check command must output metrics in the supported metrics formats.

Reviewer: ```suggestion
In addition, the check command output must present metrics in one of the supported metrics formats.
```

---------------------|------
output_metric_format | `graphite_plaintext`
documentation        | [Graphite Plaintext Protocol][14]
example              | {{< code plain >}}local.random.diceroll 4 123456789{{< /code >}}

| InfluxDB           |      |
---------------------|------
output_metric_format | `influxdb_line`
documentation        | [InfluxDB Line Protocol][15]
example              | {{< code plain >}}weather,location=us-midwest temperature=82 1465839830100400200{{< /code >}}

| OpenTSDB           |      |
---------------------|------
output_metric_format | `opentsdb_line`
documentation        | [OpenTSDB Data Specification][16]
example              | {{< code plain >}}sys.cpu.user 1356998400 42.5 host=webserver01 cpu=0{{< /code >}}

| Prometheus         |      |
---------------------|------
output_metric_format | `prometheus_text`
documentation        | [Prometheus Exposition Text][17]
example              | {{< code plain >}}http_requests_total{method="post",code="200"} 1027 1395066363000{{< /code >}}

## Enrich metrics with tags

[Output metric tags][1] are custom tags you can apply to enrich the metric points produced by check output metric extraction.

> 

Author: I think something is missing here: need to explain where tag values come from if you *don't* use token substitution.

Reviewer: I believe this depends on the metric format. Graphite plaintext, as an example, doesn't support tags. I suggest asking engineering to comment on which output metric formats are expected to produce tags.

Author: I found it in the sensu-go issue: https://github.com/sensu/sensu-go/issues/2160#issuecomment-430678869

The events reference includes an [example event with check and metric data][20].

You do not need to add a mutator to your check definition to process metrics with an event handler.
The [metrics attribute][5] format automatically reduces metrics data complexity so event handlers can process metrics effectively.

## Example metric check

This check definition collects metrics in Nagios Performance Data [format][9], enriches the metrics with [custom tags][1], and sends the collected and tagged metrics to the [Prometheus Pushgateway Handler][19] for long-term storage and visualization.

{{< language-toggle >}}

{{< code yml >}}
type: CheckConfig
api_version: core/v2
metadata:
  annotations:
  - slack-channel: '#monitoring'
  labels:
  - region: us-west-1
  name: collect-metrics
  namespace: default
spec:
  check_hooks: null
  command: collect.sh
  discard_output: true
  env_vars: null
  handlers: []
  high_flap_threshold: 0
  interval: 10
  low_flap_threshold: 0
  output_metric_format: nagios_perfdata
  output_metric_handlers:
  - prometheus_gateway
  output_metric_tags:
  - name: response_time_in_ms

>     value: '{{ .time }}'

Author: I'm not sure if this tag is a good example for Nagios format --> the Prometheus Pushgateway handler.

Also, I think I need to add another tag example that doesn't use token substitution.

Reviewer: I agree that we should test this example to see if it works out the way we think. 👍

example              | {{< code plain >}}sys.cpu.user 1356998400 42.5 host=webserver01 cpu=0{{< /code >}}

| Prometheus         |      |
---------------------|------
output_metric_format | `prometheus_text`
documentation        | [Prometheus Exposition Text][17]
example              | {{< code plain >}}http_requests_total{method="post",code="200"} 1027 1395066363000{{< /code >}}

## Enrich metrics with tags

[Output metric tags][1] are custom tags you can apply to enrich the metric points produced by check output metric extraction.

Use output metric tags for the output metric formats that do not natively support tags: Graphite Plaintext Protocol and Nagios Performance Data.

Values for output metric tags are passed through to the metric points produced by check output metric extraction for formats that natively support tags (InfluxDB Line Protocol, OpenTSDB Data Specification, and Prometheus Exposition Text).

You can use [check token substitution][22] for the [value attribute][21] to include any event attribute in an output metric tag.
For example, this tag will list the `event.time` attribute:

{{< code shell >}}
"output_metric_tags": [
  {
    "name": "instance",

>     "value": "{{ .time }}"

Reviewer 2: This example could be better.  I'd use something like `{{.entity.system.hostname}}` if you're going to have `instance` as the tag name.

  }
]{{< /code >}}

## Process extracted and tagged metrics

Specify the event handlers you want to process your Sensu metrics in the check [`output_metric_handlers`][3] attribute.
With these event handlers, you can route metrics to one or more databases for storing and visualizing metrics, like Elasticsearch, InfluxDB, Grafana, and Graphite.

Many of our most popular metrics integrations for [time-series and long-term event storage][18] include curated, configurable quick-start templates to integrate Sensu with your existing workflows.
You can also use [Bonsai][8], the Sensu asset hub, to discover, download, and share dynamic runtime assets for processing metrics.

> In check definitions, the `output_metrics_handlers` list for metrics event handlers is distinct and separate from the `handlers` list for status event handlers.

Reviewer 2: this should be `output_metric_handlers`
